### PR TITLE
Additionaltests

### PR DESCRIPTION
**Pull request by dvdstelt**
connects to #3 

- Saga initiation
- Saga retrieval
- SendLocal

### CODE REVIEW

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs [hunk 1]
```diff
@@ -48,20 +56,27 @@ public virtual void Execute(Permutation permutation, string endpointName)
...
 #if Version5
-            using(EndpointInstance){}
+            using (EndpointInstance) { }
 #else
             EndpointInstance.Stop().GetAwaiter().GetResult();
 #endif
         }
     }
 
+    void ThrowIfPermutationNotAllowed()
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: Why it the scenario testing if a permutation is allowed?

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs [hunk 2]
```diff
@@ -109,17 +124,33 @@ ISendOnlyBus CreateSendOnlyEndpoint()
...
+        if (this is ICreateSeedData) return true;
+        return false;
+    }
+
+    BusConfiguration CreateConfiguration()
     {
         var configuration = new Configuration();
         configuration.EndpointName(EndpointName);
         configuration.EnableInstallers();
+        configuration.DiscardFailedMessagesInsteadOfSendingToErrorQueue();
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: Don't remove `DiscardFailedMessagesInsteadOfSendingToErrorQueue`. This should not be disabled for *all* test.

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs [hunk 3]
```diff
@@ -153,12 +184,50 @@ Configuration CreateConfiguration()
...
 
         return configuration;
     }
 #endif
 
+    List<Type> GetTypesToInclude()
+    {
+        var location = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+        var asm = new NServiceBus.Hosting.Helpers.AssemblyScanner(location).GetScannableAssemblies();
+        //var asm = AssemblyScanner.GetAssemblies().Distinct();
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: Remove commented stuff

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs [hunk 4]
```diff
@@ -153,12 +184,50 @@ Configuration CreateConfiguration()
...
+
+        var allTypes = (from a in asm.Assemblies
+                        from b in a.GetLoadableTypes()
+                        select b).Distinct().ToList();
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: Why does this need a distinct?

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs [hunk 5]
```diff
@@ -153,12 +184,50 @@ Configuration CreateConfiguration()
...
+        return GetTypesToExclude(Assembly.GetAssembly(this.GetType()).GetTypes());
+    } 
+
+    private IEnumerable<Type> GetTypesToExclude(IEnumerable<Type> allTypes)
+    {
+        var allTypesToExclude = (from t in allTypes
+            where (t.IsSubclassOf(typeof(BaseRunner)) || t.IsSubclassOf(typeof(LoopRunner)) || t == typeof(LoopRunner.Handler)) && t != this.GetType()
+            select t).ToList();
+
+        Log.Info($"This is test {this.GetType().Name}, excluding :");
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: No string interpolation when logging.

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs [hunk 6]
```diff
@@ -177,5 +246,19 @@ public T GetConfiguration<T>() where T : class, new()
 
         return ConfigurationManager.GetSection(typeof(T).Name) as T;
     }
+}
 
+public static class Whatever
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: Whatever?

File: src/PerformanceTests/Common/Scenarios/GatedPublish/GatedPublishRunner.V5.cs [hunk 7]
```diff
@@ -4,22 +4,9 @@
...
-        EndpointInstance.Publish(msg);
-    }
-
-    public class Handler : IHandleMessages<Event>
-    {
-        public void Handle(Event message)
-        {
-            X.Signal();
-        }
+        await Task.Run(() => EndpointInstance.Publish(message));
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/GatedPublish/GatedPublishRunner.V5.cs]: Why Task.Run?

File: src/PerformanceTests/Common/Scenarios/GatedPublish/GatedPublishRunner.cs [hunk 8]
```diff
@@ -72,7 +27,7 @@ public MessageEndpointMappingCollection GenerateMappings()
 
         var messageType = typeof(Event);
 
-        Log.InfoFormat("Mapping {0} to {1}", messageType, EndpointName);
+        Log.Info($"Mapping {messageType} to {EndpointName}");
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/GatedPublish/GatedPublishRunner.cs]: No string interpolation when logging.

File: src/PerformanceTests/Common/Scenarios/LoopRunner.cs [hunk 9]
```diff
@@ -1,14 +1,29 @@
...
 
 abstract class LoopRunner : BaseRunner
 {
+    ILog Log = LogManager.GetLogger(typeof(LoopRunner));
+
     Task loopTask;
-    protected CancellationTokenSource stopLoop { get; private set; }
-    protected bool Shutdown { get; private set; }
+
+    static CountdownEvent countdownEvent { get; set; }
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/LoopRunner.cs]: This is part of the scenario instance not of its base class.

File: src/PerformanceTests/Common/Scenarios/LoopRunner.cs [hunk 10]
```diff
@@ -1,14 +1,29 @@
...
+
+    CancellationTokenSource stopLoop { get; set; }
+    bool Shutdown { get; set; }
+    protected abstract int BatchSize { get; set; }
+    protected abstract Task SendMessage();
 
     protected override void Start()
     {
+        if (BatchSize == default(int))
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/LoopRunner.cs]: Why should this be set? What is a scenario doesn't do anything batch oriented?

File: src/PerformanceTests/Common/Scenarios/LoopRunner.cs [hunk 11]
```diff
@@ -26,5 +41,73 @@ protected override void Stop()
...
 
-    protected abstract Task Loop(object o);
+    public class Handler : IHandleMessages<IMessage>
+    {
+        ILog Log = LogManager.GetLogger(typeof(Handler));
+
+#if Version5
+        public void Handle(IMessage message)
+        {
+            Log.Info($"This is {this.GetType()}");
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/LoopRunner.cs]: No string interpolation when logging.

File: src/PerformanceTests/Common/Scenarios/Receive/ReceiveRunner.V5.cs [hunk 12]
```diff
@@ -1,12 +1,13 @@
...
+using Common.Scenarios;
 using NServiceBus;
 
 /// <summary>
 /// Does a continuous test where a pre-seeded amount of messages will be handled
 /// </summary>    
 partial class ReceiveRunner : ICreateSeedData
 {
-    public int SeedSize { get; set; } = 50000;
+    public int SeedSize { get; set; } = 25000;
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/Receive/ReceiveRunner.V5.cs]: Why 25,000?

File: src/PerformanceTests/Common/Scenarios/SagaInitiateRunner.V5.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        if (address == null)
+        {
+            var unicastBus = (NServiceBus.Unicast.UnicastBus)endpointInstance;
+            var machine = unicastBus.Configure.LocalAddress.Machine;
+            var queue = unicastBus.Configure.LocalAddress.Queue;
+
+            address = new Address(queue, machine);
+        }
+
+        endpointInstance.Send(address, new Command(Interlocked.Increment(ref messageId)));
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/SagaInitiateRunner.V5.cs]: Why not `SendLocal`?

File: src/PerformanceTests/Common/Scenarios/SendRunner.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+    protected override void Start()
+    {
+    }
+
+    protected override void Stop()
+    {
+    }
+
+    public class Command : ICommand
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/SendRunner.cs]: Why does `LoopRunner` have a `Command`?

File: src/PerformanceTests/Common/Scenarios/SendRunner.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+/// <summary>
+/// Does a continuous test where a pre-seeded amount of messages will be handled
+/// </summary>    
+class SendRunner : LoopRunner
+{
+    protected override int BatchSize { get; set; } = 16;
+
+    protected override Task SendMessage()
+    {
+        throw new System.NotImplementedException();
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/SendRunner.cs]: Why must it override to the throw a `NotImplementedException`? This identifies that there is something wrong with the base class.

File: src/PerformanceTests/Tests/Categories/PublishVsSendFixture.cs [hunk 16]
```diff
@@ -0,0 +1,49 @@
+namespace Categories
+{
+    using System.Collections.Generic;
+    using NUnit.Framework;
+    using Tests.Permutations;
+    using Variables;
+
+    [TestFixture(Description = "Publish vs Send", Category = "Performance")]
+    public class PublishVsSendFixture : Base
```
Comment (ramonsmits) [inline src/PerformanceTests/Tests/Categories/PublishVsSendFixture.cs]: Why is this recreated? The `PublishVsSendFixture` was renamed to `TransportsFixture`.

File: src/PerformanceTests/Transport.V5.AzureStorageQueues.V6/AzureStorageQueueProfile.cs [hunk 17]
```diff
@@ -7,5 +8,7 @@ public void Configure(BusConfiguration busConfiguration)
         busConfiguration
             .UseTransport<AzureStorageQueueTransport>()
             .ConnectionString(this.GetConnectionString("AzureStorageQueue"));
+
+        //ServicePointManager.UseNagleAlgorithm = false;
```
Comment (ramonsmits) [inline src/PerformanceTests/Transport.V5.AzureStorageQueues.V6/AzureStorageQueueProfile.cs]: Remove commented code.

File: src/PerformanceTests/Common/Common.Data.projitems [hunk 18]
```diff
@@ -17,17 +17,25 @@
...
     <Compile Include="$(MSBuildThisFileDirectory)Scenarios\GatedSendLocal\GatedSendLocalRunner.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Scenarios\GatedSendLocal\GatedSendLocalRunner.V5.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Scenarios\GatedSendLocal\GatedSendLocalRunner.V6.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Scenarios\IThrowIfPermutationIsNotAllowed.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Scenarios\ICreateSeedData.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Scenarios\Receive\ReceiveRunner.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Scenarios\Receive\ReceiveRunner.V5.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Scenarios\Receive\ReceiveRunner.V6.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Scenarios\SendRunner.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Scenarios\SagaRetrievalRunner.V6.cs" />
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Common.Data.projitems]: Group files that belong together in a subfolder.

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs [hunk 19]
```diff
@@ -153,12 +184,50 @@ Configuration CreateConfiguration()
...
+        var location = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+        var asm = new NServiceBus.Hosting.Helpers.AssemblyScanner(location).GetScannableAssemblies();
+        //var asm = AssemblyScanner.GetAssemblies().Distinct();
+
+        var allTypes = (from a in asm.Assemblies
+                        from b in a.GetLoadableTypes()
+                        select b).Distinct().ToList();
+
+        IEnumerable<Type> allTypesToExclude = GetTypesToExclude(allTypes);
+        var finalInternalListToScan = allTypes.Except(allTypesToExclude);
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: Nice trick to make that work with V6!

File: src/PerformanceTests/Common/Scenarios/SagaRetrievalRunner.cs [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+
+partial class SagaRetrievalRunner : BaseRunner, IThrowIfPermutationIsNotAllowed
+{
+    public void ThrowIfPermutationIsNotAllowed(Permutation permutation)
+    {
+        if (permutation.Persister == Persistence.InMemory)
+            throw new InvalidOperationException("This test has no use running with InMemory persistence.");
+    }
+
+    protected override void Start()
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/SagaRetrievalRunner.cs]: Weird to having to override methods without implementing anything.

File: src/PerformanceTests/Common/Scenarios/SagaRetrievalRunner.cs [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    {
+    }
+
+    protected override void Stop()
+    {
+    }
+
+    public class Command : ICommand
+    {
+        public Command(int id)
```
Comment (ramonsmits) [inline src/PerformanceTests/Common/Scenarios/SagaRetrievalRunner.cs]: Did you test this with the serializers? i would not assume that all serializers can deserialize a message with constructor that has args.

File: src/PerformanceTests/Tests/Categories/MsmqFixture.cs [hunk 22]
```diff
@@ -15,6 +15,12 @@ public override void ReceiveRunner(Permutation permutation)
         }
 
         [TestCaseSource(nameof(CreatePermutations))]
+        public override void SagaInitiateRunner(Permutation permutation)
```
Comment (ramonsmits) [inline src/PerformanceTests/Tests/Categories/MsmqFixture.cs]: Why would this be added to the MSMQ fixture?

File: src/PerformanceTests/Tests/Categories/PersistersFixture.cs [hunk 23]
```diff
@@ -15,6 +15,12 @@ public override void GatedPublishRunner(Permutation permutation)
             base.GatedPublishRunner(permutation);
         }
 
+        [TestCaseSource(nameof(CreatePermutations))]
+        public override void SagaInitiateRunner(Permutation permutation)
+        {
```
Comment (ramonsmits) [inline src/PerformanceTests/Tests/Categories/PersistersFixture.cs]: Adding it only here would make sense.

**Comment by ramonsmits** [comment]
I'm not a fan of the BaseRunner refactorings. It makes the scenarios very brittle. When the base class changes, the scenarios must change too. 

It also seems to enforce only using a single message due to the signature of `SendMessage()`.

I would not be in favor of merging this with the current refactorings.

File: src/PerformanceTests/Tests/Categories/PublishVsSendFixture.cs (hunk 16 above)
Comment (dvdstelt) [inline src/PerformanceTests/Tests/Categories/PublishVsSendFixture.cs]: Apparently not deleted from my file system on rebase. Removed it from file system.

File: src/PerformanceTests/Common/Scenarios/SendRunner.cs (hunk 14 above)
Comment (dvdstelt) [inline src/PerformanceTests/Common/Scenarios/SendRunner.cs]: It doesn't?

File: src/PerformanceTests/Common/Scenarios/SagaInitiateRunner.V5.cs (hunk 13 above)
Comment (dvdstelt) [inline src/PerformanceTests/Common/Scenarios/SagaInitiateRunner.V5.cs]: SendLocal isn't available with an ISendOnlyBus

File: src/PerformanceTests/Common/Scenarios/Receive/ReceiveRunner.V5.cs (hunk 12 above)
Comment (dvdstelt) [inline src/PerformanceTests/Common/Scenarios/Receive/ReceiveRunner.V5.cs]: Why 50k? Why 10k? This is outside of the scope of this PR.

File: src/PerformanceTests/Common/Scenarios/LoopRunner.cs (hunk 10 above)
Comment (dvdstelt) [inline src/PerformanceTests/Common/Scenarios/LoopRunner.cs]: Then it should not use LoopRunner but BaseRunner.

File: src/PerformanceTests/Common/Scenarios/BaseRunner.cs (hunk 1 above)
Comment (dvdstelt) [inline src/PerformanceTests/Common/Scenarios/BaseRunner.cs]: Because running a Saga with InMemory persistence and restarting the endpoint, makes everything disappear and the test worthless. I thought of not throwing an exception, but just ignoring the test. But this was something I wanted to ask, before merging.

**Comment by dvdstelt** [comment]
Aren't you mixing LoopRunner and BaseRunner? I also have the idea you did so in the comments. Only thing that was actually changed in BaseRunner is the 'scanning' part.

Initially there was *a lot* of duplication in tests. I first introduced BaseRunner in a previous PR that supported a few scenarios. LoopRunner already supported two scenarios and I added a third to it. Saw about 30 to 40 lines of code duplication (3 times) so I merged that into the already existing LoopRunner. It now supports 3 specific scenarios, with BaseRunner being more generic and covering the other scenarios. I can't look into the future, but if a scenario will be changed (or added), it might be extracted from the current base class and possible get is own base class, or just completely custom code. But as long as they can share the same code...

Only thing that's not solved yet is the static countdownEvent object. On that you're right. In the future, if we'll run several tests at the same time in stability tests, that will conflict and needs to be resolved.

**Comment by dvdstelt** [comment]
@ramonsmits Is it an idea to merge this and fix the static countDownEvent in another PR? Helps with rebasing/merging and I feel we're currently doing stuff double in other PRs.
